### PR TITLE
[DCJ-32] Load tags are locked and unlocked at the dataset level

### PR DESCRIPTION
**Pull request by okotsopoulos**
__Jira ticket__: https://broadworkbench.atlassian.net/browse/DCJ-32

## Addresses

When files are being ingested to a dataset (via standalone file ingest or as part of a combined metadata-and-file ingest), TDR attempts to take out a lock on a **load tag**, either provided by the user or generated with the form `load-at-<timestamp>`.  (This is in addition to the dataset-level lock taken out during the duration of the ingest.)

A user may elect to provide and reuse their own load tag to logically link file loads together, but TDR will not allow a load tag to be used by multiple running file ingests: it confuses the algorithm for re-running a load with a load tag and skipping already-loaded files.

## Summary of changes

Let `flight1` have an active lock on `loadTag` while ingesting files to a dataset.

| `flight2` attempts to lock `loadTag`… | Previous Behavior | New Behavior |
|--------|--------|--------|
| …while ingesting files to the **same dataset** | ❌ | ❌ |
| …while ingesting files to a **different dataset** | ❌ | ✅ |

Note that this PR is necessary for us to expand our [unlockDataset](https://data.terra.bio/swagger-ui.html#/datasets/unlockDataset) endpoint to also clear any stuck locks on load tags.  We recently had to modify the database directly to clear such stuck locks: https://broadworkbench.atlassian.net/browse/DCJ-625

## Testing Strategy

Expanded unit tests (with drive-bys for JUnit 5 conversion, downgrading a Spring Boot test with embedded database to a pure unit test with mocking).

### CODE REVIEW

File: src/main/java/bio/terra/service/load/Load.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/load/Load.java:1]: Drive-by: conversion to Java Record.

File: src/main/java/bio/terra/service/load/flight/LoadLockStep.java [hunk 1]
```diff
@@ -9,21 +9,24 @@
...
-
 public class LoadLockStep implements Step {
   private final LoadService loadService;
+  private final UUID datasetId;
 
-  public LoadLockStep(LoadService loadService) {
+  /**
+   * This step is meant to be shared by dataset and filesystem flights for locking the load tag. It
+   * expects to find LoadMapKeys.LOAD_TAG in the parameters or working map.
+   */
```
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/load/flight/LoadLockStep.java]: Top-level comments are more useful as constructor Javadoc -- we construct new Steps when we define our flights, so you're more likely to see this context in the flight that way in your IDE.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java:1]: Conversion to JUnit 5.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 2]
```diff
@@ -1,51 +1,57 @@
...
+import org.springframework.test.context.junit.jupiter.SpringExtension;
 
-@RunWith(SpringRunner.class)
+@ExtendWith(SpringExtension.class)
 @SpringBootTest
 @AutoConfigureMockMvc
 @ActiveProfiles({"google", "unittest"})
-@Category(Unit.class)
+@Tag(Unit.TAG)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Needed for `@BeforeAll` and `@AfterAll` annotations on non-static methods.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 3]
```diff
@@ -80,8 +87,17 @@ public String getId() {
...
 
+  private List<Dataset> datasets;
   private List<UUID> loadIdsWithFilesUsedByTest;
 
+  @BeforeAll
+  void beforeAll() throws IOException {
+    datasets =
+        List.of(
+            daoOperations.createDataset(DaoOperations.DATASET_MINIMAL),
+            daoOperations.createDataset(DaoOperations.DATASET_MINIMAL));
```
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: I added a foreign key to our `load` table, so adding new `load` records with dataset IDs will require a corresponding `dataset` record.  That necessitated an expansion to our DB test set-up, but we can set up these records once for the whole class rather than for every test invocation.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 4]
```diff
@@ -188,78 +208,60 @@ public void loadFilesTest() throws Exception {
...
-    try {
-      loadDao.lockLoad(loadTag, flightId);
-      return true;
-    } catch (LoadLockedException ex) {
-      return false;
-    }
-  }
-
-  @Test
-  public void conflictLockTest() throws Exception {
```
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java:226]: Test deleted because it asserted nothing.

File: src/test/java/bio/terra/service/load/LoadLockTester.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadLockTester.java:1]: Deleted: unused.

File: src/test/java/bio/terra/service/load/LoadLockUnlockLooper.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadLockUnlockLooper.java:1]: Deleted: only used in deleted test.

File: src/test/java/bio/terra/service/load/LoadServiceTest.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadServiceTest.java:1]: Conversion to JUnit 5 and downgrade to strict unit test with mocking.  Otherwise, this exercises `LoadDao` rather than `LoadService`.

File: src/test/java/bio/terra/service/load/LoadLockedBy.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/load/LoadLockedBy.java:1]: An unexpected side effect of converting `Load` to a record is that I wasn't able to use Hamcrest `hasProperty` matchers in assertions: https://stackoverflow.com/questions/66982522/how-can-i-assert-hasproperty-with-a-java-record

So, I built a custom matcher to simplify assertions on a load tag's lock.

File: src/main/java/bio/terra/service/load/flight/LoadLockStep.java [hunk 5]
```diff
@@ -9,21 +9,24 @@
...
+  public LoadLockStep(LoadService loadService, UUID datasetId) {
     this.loadService = loadService;
+    this.datasetId = datasetId;
```
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/load/flight/LoadLockStep.java]: If reviewers prefer, I could instead add an `IngestUtils` method to pull the `datasetId` from the flight's input parameters, rather than pass it into step construction.  We are not consistent on approach here.

Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/load/flight/LoadLockStep.java]: That's what I was going to suggest, the dataset ID is always in `JobMapKeys.DATASET_ID` in these flights so I'd extract it in the step instead of passing it down via constructor.

Comment (okotsopoulos) [reply src/main/java/bio/terra/service/load/flight/LoadLockStep.java]: Updated!

File: src/main/java/bio/terra/service/load/Load.java [hunk 6]
```diff
@@ -1,58 +1,22 @@
...
+    String lockingFlightId,
+    UUID datasetId,
+    Instant createdDate) {
+
+  /**
+   * @return whether this Load is locked by the supplied flight ID and dataset
+   */
+  public boolean lockedBy(String lockingFlightId, UUID datasetId) {
+    return Objects.equals(this.lockingFlightId, lockingFlightId)
+        && Objects.equals(this.datasetId, datasetId);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/load/Load.java]: Is it possible for `lockingFlightId` or `datasetId` to be `null`? If not, this is the same as calling `equals()`.
```suggestion
    return lockingFlightId.equals(lockingFlightId, lockingFlightId)
        && datasetId.equals(datasetId, datasetId);
```

Using `equals()` is slightly shorter and makes it clearer that `null`s are not expected here.

File: src/test/java/bio/terra/service/load/flight/LoadLockStepTest.java [hunk 7]
```diff
@@ -72,11 +79,21 @@ void doStep_retryLoadLockedException() throws InterruptedException {
...
 
     assertThat(workingMap.get(LoadMapKeys.LOAD_ID, UUID.class), equalTo(null));
   }
+
+  @Test
+  void undoStep() {
+    StepResult undoResult = step.undoStep(flightContext);
+
+    assertThat(undoResult.getStepStatus(), equalTo(StepStatus.STEP_RESULT_SUCCESS));
+    assertThat(undoResult.getException().isPresent(), equalTo(false));
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/load/flight/LoadLockStepTest.java]: What is the purpose of checking that the exception is missing? Is there a code path where exception is set and the flight status is `SUCCESS`?

Also, since `StepResult.getStepResultSuccess()` returns a static object, it can be compared directly:
```suggestion
    assertThat(undoResult, equalTo(StepResult.getStepResultSuccess()));
```

File: src/main/java/bio/terra/service/load/Load.java [hunk 8]
```diff
@@ -1,58 +1,22 @@
...
+public record Load(
+    UUID id,
+    String loadTag,
+    boolean locked,
+    String lockingFlightId,
+    UUID datasetId,
+    Instant createdDate) {
+
+  /**
+   * @return whether this Load is locked by the supplied flight ID and dataset
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/load/Load.java]: It's a minor semantic issue but for a `Load` object this seems more like `isSameLockAs()` since a load itself is not locked, the dataset is locked.

File: src/main/java/bio/terra/service/load/Load.java [hunk 9]
```diff
@@ -1,58 +1,22 @@
...
+   */
+  public boolean lockedBy(String lockingFlightId, UUID datasetId) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/load/Load.java]: We're not consistent here but I think functions that return `boolean` should in general start with `is` or `has`
```suggestion
  public boolean isLockedBy(String lockingFlightId, UUID datasetId) {
```

File: src/main/java/bio/terra/service/load/Load.java [hunk 10]
```diff
@@ -1,58 +1,22 @@
...
-    return new ToStringBuilder(this, ToStringStyle.JSON_STYLE)
-        .append("id", id)
-        .append("loadTag", loadTag)
-        .append("locked", locked)
-        .append("lockingFlightId", lockingFlightId)
-        .toString();
+public record Load(
+    UUID id,
+    String loadTag,
+    boolean locked,
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/load/Load.java]: The code stores `locked` but I don't see it used anywhere. Is it needed?

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java (hunk 3 above)
Comment (pshapiro4broad) [reply src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: What is the time tradeoff of using `BeforeAll` vs `BeforeEach`? All other tests that use `daoOperations.createDataset()` are using `BeforeEach`. I would merge this code with `BeforeEach` and see if it performs the same.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 11]
```diff
@@ -125,8 +140,13 @@ public void teardown() {
     }
   }
 
+  @AfterAll
+  void afterAll() {
+    datasets.forEach(daoOperations::deleteDatasetCascade);
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Since this test is using the embedded database, I would expect it to be cleaned up after the test is complete. I would delete both `afterAl()` and `afterEach()` and see if things still work, as I think the current code is doing extra steps for no benefit.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 12]
```diff
@@ -1,51 +1,57 @@
...
-
   private enum LoadTagsUsedByTest {
     LOADTAG_MY("myLoadTag"),
     LOADTAG_SERIAL("serialLoadTag"),
     LOADTAG_CONCURRENT("concurrentLoadTag"),
     LOADTAG_CONFLICT("conflictLoadTag");
-    private String tag;
+    private final String tag;
 
     public String getTag() {
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Note that since this is a `private` inner class (enum) there's no point in having getters as the enclosing class can access the enum's fields directly.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 13]
```diff
@@ -80,8 +87,17 @@ public String getId() {
...
   }
 
+  private List<Dataset> datasets;
   private List<UUID> loadIdsWithFilesUsedByTest;
 
+  @BeforeAll
+  void beforeAll() throws IOException {
+    datasets =
+        List.of(
+            daoOperations.createDataset(DaoOperations.DATASET_MINIMAL),
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Now there are a few calls to `daoOperations.createDataset()` with `DaoOperations.DATASET_MINIMAL`, maybe we should have an overloaded version with no args that uses this value?

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 14]
```diff
@@ -90,16 +106,15 @@ public String getId() {
...
-      // loop through all flight ids in the enum, since any one could have successfully locked the
-      // load last
+      // loop through all combos of flight id enum values and datasets, since any one could have
+      // successfully locked the load last
       for (FlightIdsUsedByTest flightId : FlightIdsUsedByTest.values()) {
-        try {
-          loadDao.unlockLoad(loadTag.getTag(), flightId.getId());
-        } catch (RuntimeException rEx) {
+        for (Dataset dataset : datasets) {
+          loadDao.unlockLoad(loadTag.getTag(), flightId.getId(), dataset.getId());
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: I think this is also unnecessary, as the database is emptied

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java [hunk 15]
```diff
@@ -188,78 +208,60 @@ public void loadFilesTest() throws Exception {
...
-    assertFalse("y does not get lock", ylocks);
-
-    loadDao.unlockLoad(loadTag, flightX);
-
-    ylocks = tryLockLoad(loadTag, flightY);
-    assertTrue("y gets lock", ylocks);
+    assertThat(
+        "[Flight=X,dataset=A] gets lock",
+        loadDao.lockLoad(loadTag, flightX, datasetIdA),
+        loadLockedBy(flightX, datasetIdA));
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Is a custom matcher needed here? I think this is the same as
```suggestion
    assertThat(
        "[Flight=X,dataset=A] gets lock",
        loadDao.lockLoad(loadTag, flightX, datasetIdA)
            .lockedBy(flightX, datasetIdA));
```

File: src/main/java/bio/terra/service/load/LoadDao.java [hunk 16]
```diff
@@ -72,83 +78,94 @@ public LoadDao(NamedParameterJdbcTemplate jdbcTemplate) {
...
-            "Load '" + loadTag + "' is locked by flight '" + load.getLockingFlightId() + "'");
-      }
-    } else {
-      load =
-          new Load()
-              .id(keyHolder.getId())
-              .loadTag(keyHolder.getString("load_tag"))
-              .locked(keyHolder.getField("locked", Boolean.class).orElse(false))
-              .lockingFlightId(keyHolder.getString("locking_flight_id"));
+    jdbcTemplate.update(upsert, params, keyHolder);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/load/LoadDao.java]: Since `keyHolder` is now unused, it can be removed.

Or, you can preserve the existing logic by building a `Load` object if the `update()` command returns a nonzero value. The only field besides the key that's db generated is the date, which could be generated in this method instead of in the database.

File: src/main/java/bio/terra/service/load/LoadDao.java [hunk 17]
```diff
@@ -72,83 +78,94 @@ public LoadDao(NamedParameterJdbcTemplate jdbcTemplate) {
...
+    Load load = lookupLoad(loadTag, datasetId);
+    if (load == null) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/load/LoadDao.java]: Was there a reason why this logic was changed? I like the original in that I can see that it only loads the record if a conflict occurs, and it only checks for an existing lock in that case as well. The new code is more confusing to me; why would the record we just wrote be missing? Why would we check `lockedBy()` if update was successful?

File: src/main/java/bio/terra/service/load/LoadDao.java [hunk 18]
```diff
@@ -72,83 +78,94 @@ public LoadDao(NamedParameterJdbcTemplate jdbcTemplate) {
    *
    * @param loadTag tag identifying this load
    * @param flightId flight id taking the lock
+   * @param datasetId dataset to which this load is being applied
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/load/LoadDao.java]: I think the javadoc (the general text above) can be further updated to describe how `datasetId` is used

**Review by pshapiro4broad — APPROVED** [review]
Many comments, but mostly minor things

- I would like to see the DAO test cleaned up a bit since I think teardown and some prep are not needed with the embedded database
- If `Load.locked` is unused, it would be nice to remove it

File: src/test/java/bio/terra/service/load/flight/LoadLockStepTest.java (hunk 7 above)
Comment (okotsopoulos) [reply src/test/java/bio/terra/service/load/flight/LoadLockStepTest.java]: It's technically possible to set an exception for a successful result, but realistically it's unlikely to happen.  Your suggested change is a good one, it sidesteps this entirely.

File: src/main/java/bio/terra/service/load/Load.java (hunk 8 above)
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/load/Load.java]: We are actually locking the load tag in addition to the dataset lock.  Does that change your perspective here or am I missing something?

File: src/main/java/bio/terra/service/load/Load.java (hunk 10 above)
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/load/Load.java]: No, I don't think it is. I'll at least remove it from our code but will hold off on dropping the column for now.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java (hunk 11 above)
Comment (okotsopoulos) [reply src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Interesting, I'll experiment with this.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java (hunk 14 above)
Comment (okotsopoulos) [reply src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: This one I'm less convinced of (mostly existing code), if different test cases are reusing the same load tags and flights without fully cleaning them up.  But I am not well-versed in what the embedded database's behavior is between test invocations.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java (hunk 15 above)
Comment (okotsopoulos) [reply src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: This would be an equivalent assertion, but I don't like asserting a truth-y condition on a complex object if I can help it… if the assertion fails, we're lacking in helpful debugging info in the error message rather than seeing the full Load object failing the condition.

But, a custom matcher does add complexity, so I don't mind going to a truth-y assertion if that's what reviewers prefer.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java (hunk 14 above)
Comment (okotsopoulos) [reply src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Looking at the logs, I'm seeing that we spin up a DB at the start of the test suite and initiate connections to it for every invocation.  So I think we could end up in a scenario where later test method invocations are working off of a dirty DB.

That being said… by moving dataset construction to `@BeforeEach` we are going to get a unique dataset ID for each test method invocation.  Any preexisting load locks would be for different datasets and not conflict.

Similarly, the existing `@AfterEach` method is running `loadDao.cleanFiles(loadId)` for all `loadId`s recorded… but these load IDs are UUIDs generated by the DB for each load tag lock, so I don't think those pose a conflict either. 

I'd be interested to hear your thoughts on this and whether I'm getting warmer in my understanding of this behavior.

Comment (okotsopoulos) [reply src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Last update for now -- I gave this test class a big face lift.

**Review by fboulnois — APPROVED** [review]
looks good 👍

**Review by rushtong — APPROVED** [review]
Looks good, thank you 👍🏽

File: src/main/java/bio/terra/service/load/Load.java (hunk 8 above)
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/load/Load.java]: Interesting, I didn't realize that the load tag itself was locked, but I can see how that makes sense. I think what threw me here is that `Load` has a property `locked` but doesn't use it to return if it's locked.

Maybe the thing I'm missed is that `Load` is really `LoadTag`? If so, renaming the class may be out of scope but a javadoc with that info could help.

Either way I'm OK with leaving the method name as-is if it makes sense to you that way

Comment (okotsopoulos) [reply src/main/java/bio/terra/service/load/Load.java]: > I think what threw me here is that Load has a property locked but doesn't use it to return if it's locked

Yes!  I definitely share in that confusion.

>Maybe the thing I'm missed is that Load is really LoadTag? If so, renaming the class may be out of scope but a javadoc with that info could help.

That's a good idea -- I'll see about a rename.

File: src/main/java/bio/terra/service/load/LoadDao.java (hunk 18 above)
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/load/LoadDao.java]: Good shout -- updated.

File: src/main/java/bio/terra/service/load/LoadDao.java (hunk 17 above)
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/load/LoadDao.java]: I mainly didn't care for the near-duplication of load construction: we were retrieving the load either way.  But these are good points, I've revisited the code and believe I've addressed your concerns while avoiding the duplicative construction.

File: src/main/java/bio/terra/service/load/Load.java (hunk 8 above)
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/load/Load.java]: I spent a fair bit of time thinking about this one and had an "aha" moment.  It's really that `Load` is more like `LoadLock` -- the load key that's being locked (previously, just the load tag and now the combo of load tag and dataset ID) along with the flight ID doing the locking.

The method that I added here wasn't actually necessary: the prior approach of checking for a flight ID match still holds today.  But I renamed to `LoadLock` and restructured to make this behavior clearer, I hope.

File: src/test/java/bio/terra/service/load/LoadDaoUnitTest.java (hunk 15 above)
Comment (okotsopoulos) [reply src/test/java/bio/terra/service/load/LoadDaoUnitTest.java]: Update: refactoring meant that this custom matcher was no longer valuable (let alone necessary), I removed it.
